src/mlpack/methods/decision_tree/best_binary_numeric_split.hpp

```diff
@@ -63,13 +76,94 @@ class BestBinaryNumericSplit
       const WeightVecType& weights,
       const size_t minimumLeafSize,
       const double minimumGainSplit,
-      arma::vec& classProbabilities,
+      arma::vec& splitInfo,
       AuxiliarySplitInfo& aux);
 
+  /**
+   * Check if we can split a node.  If we can split a node in a way that
+   * improves on 'bestGain', then we return the improved gain.  Otherwise we
+   * return the value 'bestGain'.  If a split is made, then splitInfo and aux
+   * may be modified.
+   *
+   * This overload is used only for regression tasks.
+   *
+   * @param bestGain Best gain seen so far (we'll only split if we find gain
+   *      better than this).
+   * @param data The dimension of data points to check for a split in.
+   * @param responses Responses for each point.
+   * @param weights Weights associated with responses.
+   * @param minimumLeafSize Minimum number of points in a leaf node for
+   *      splitting.
+   * @param minimumGainSplit Minimum gain split.
+   * @param splitInfo Stores split information on a successful split.
+   * @param aux Auxiliary split information, which may be modified on a
+   *      successful split.
+   */
+  template<bool UseWeights, typename VecType, typename ResponsesType,
+           typename WeightVecType>
+  static typename std::enable_if<
+      !HasBinaryScanInitialize<FitnessFunction, void(FitnessFunction::*)
+          (const ResponsesType&, const WeightVecType&, const size_t)>::value &&
+      !HasBinaryStep<FitnessFunction, void(FitnessFunction::*)
+          (const ResponsesType&, const WeightVecType&, const size_t)>::value &&
```

Comment:
```suggestion
      !HasBinaryScanInitialize<FitnessFunction, void(FitnessFunction::*)
          (const ResponsesType&, const WeightVecType&, const size_t)>::value ||
      !HasBinaryStep<FitnessFunction, void(FitnessFunction::*)
          (const ResponsesType&, const WeightVecType&, const size_t)>::value ||
```
```suggestion
          (const ResponsesType&, const WeightVecType&, const size_t)>::value ||
```
I am not sure if this will fully fix the issue, but the correct inversion of the condition should either be !(1 && 2 && 3) or !1 || !2 || !3. :)

If that still doesn't work, I'll take a look into it tomorrow morning---it's pretty late here now, so I'm headed to bed. :)

Reply:
Okay, so I spent a while fighting with it to come to an interesting conclusion... the signatures you have appear to be correct, but `MSEGain` actually does not have a function `MSEGain::BinaryScanInitialize`... it has a function `template<bool UseWeights> MSEGain::BinaryScanInitialize<UseWeights>`!  `HAS_MEM_FUNC` can't search for this.  So I wrote a slightly different version, but it does not (quite) work and it's not very general anyway:

```
#define HAS_MEM_TEMPLATE_FUNC(FUNC, NAME)                                      \
template<typename T, typename sig, bool arg, typename = std::true_type>    \
struct NAME : std::false_type {};                                              \
                                                                               \
template<typename T, typename sig, bool arg>                               \
struct NAME                                                                    \
<                                                                              \
  T,                                                                           \
  sig,                                                                         \
  std::integral_constant<                                                      \
      bool,                                                                    \
      mlpack::sfinae::SigCheck<sig, &T::template FUNC<arg>>::value>                     \
> : std::true_type {};
```

Then, I created an auxiliary struct for nicer signatures:

```
template<typename T, bool UseWeights, typename ResponsesType, typename WeightVecType>
struct HasOptimizedBinarySplitFunctions
{
  static const bool value =
      typename HasBinaryScanInitialize<T, void(T::*)(const ResponsesType&, const
WeightVecType&, const size_t), UseWeights>::value &&
      typename HasBinaryStep<T, void(T::*)(const ResponsesType&, const WeightVecType&,
const size_t), UseWeights>::value &&
      HasBinaryGains<T, std::tuple<double, double>(T::*)()>::value;
};
```

Then in the `enable_if<>` call, you can just use `HasOptimizedBinarySplitFunctions<...>::value`.

But, as I think about it more, I am not sure it's actually worth debugging the check for *all* three functions.  I think it's fine to say, if we find `BinaryGains()`, then that's good enough to attempt to compile the optimized version.  I picked `BinaryGains()` there because it does not have a template parameter, which means that we don't have the problem above.  So, if you change the definition of `value` in the auxiliary struct to just use `HasBinaryGains<T, std::tuple<double, double>(T::*)()>::value`, it should succeed.  I think that differs from yours in that it does not have `void` for the parameter, but I am not 100% sure whether that is necessary.

Anyway, if you want to fight with the template metaprogramming more to check for all three, feel free---it could be an interesting adventure---but I think just checking for one is sufficient too. :)

Reply:
@rcurtin, It is still failing after removing HasBinaryScanInitialize and HasBinaryStep. The error is different this time. Take a look at this https://pastebin.com/gkvw3X9D.

Reply:
What have you tried to debug it?

Reply:
Check the latest commit. I simply removed the unrequired checks.

Reply:
I won't be able to work with this until tomorrow.  Do you think you can try various changes to see if they have any effect?  I was able to make the `BinaryGains` check work without problems.  Maybe removing `void` is what is needed?

Reply:
I tried all possible combinations --- removed void, moved enable_if to template parameters, tried some suggestions from StackOverflow, but none of them helped.

Reply:
I think that this commit should help: https://gist.github.com/5dfb3fd62552aa4aeb731bd382e0c6cb

I left a few other comments in the chat channel that should perhaps be helpful too:

```
I believe that the issue you are encountering is the same thing as reported here:
https://stackoverflow.com/questions/13401716/selecting-a-member-function-using-different-enable-if-conditions

basically, SFINAE for the member function of BestBinaryNumericSplit doesn't work, because we aren't actually using any template parameters specific to that member function, but instead only specific to the class
I am trying to think of a "clean" solution
```

Reply:
Thanks, @rcurtin! The fix worked like a charm :-D.